### PR TITLE
[#1873] Favourite resources

### LINKED ISSUE #1873: Logged user can mark a resource as a favourite

### LINKED ISSUE #1874: Anonymous user can mark a resource as his favourite
As in the title.
Even though the favourite functionality is dedicated for the logged user, it can be triggered by the unlogged user.
checkbox should be visible for both:

if the unlogged user clicks it for the first time during the session, pup-up appears:

https://projects.invisionapp.com/share/JXOZKC8WMDU#/screens/447696857_Favourite_Popup_Unauth
After the login in action, chosen resource should be ticked and added to his Favourite Resources list. The user stays on the resource list view (not the favourite resources), the popup from #1873 is being shown to the user as well.

Is the user choses not to log in, the box is ticked anyway. other resources can be marked as favourites as well.
Popup does not appear anymore. If the user does not log in during the session, the data is lost.
If during the session user logs in, all checked resources are added to his Favourite List

Additional context and specs:
1. text in the popup should read: 
Header:
Save your favourites!
Content:
If you want this EOSC resource to be added to your Favourite Resources list, please log in. In other case, we won’t be able to remember this resources as your favourite.
2. popups are displayed in the **bottom right corner**

### PR DESCRIPTION
**Pull request by kmarszalek**
Favourite list view
Zrzut ekranu 2021-04-14 o 10 40 01

Resource list view with checkbox
Zrzut ekranu 2021-04-14 o 10 40 23

Resource view with checkbox
Zrzut ekranu 2021-04-14 o 10 59 33

Closes #1873 
Closes #1874

### CODE REVIEW

File: app/models/service.rb [hunk 1]
```diff
@@ -78,6 +78,11 @@ class Service < ApplicationRecord
                                 reject_if: lambda { |attributes| attributes["email"].blank? },
                                 allow_destroy: true
 
+  has_many :user_service, dependent: :destroy
```
Comment (goreck888) [inline app/models/service.rb:81]: use plurals :)

**Comment by Marcelinna** [comment]
Steps to reproduce: (not logged in user)
2. Add to favotite (e.g EGI Cloud Compute) - modal "Save your favourites! " is displayed
3. Add second service to favorite (e.g EGI High-Throughput Compute)
4. Remove from favorite (EGI High-Throughput Compute)
5. Remove from favorite ( EGI Cloud Compute)
6. Login in 
7. EGI High-Throughput Compute is added to favorite dispite I deleted it
image

**Comment by Marcelinna** [comment]
https://www.awesomescreenshot.com/video/3405884?key=837e0712fc3b3f3cd2459fd6492dee2d
1. Login 
2. Add first service to favorite (modal is displayed)
3. Add other services to favorite
4. Go to Favorites resource /favourites - services are visible
5. Remove favorite from (Favorites resource)
6. Go to /services
7. Add the same favorites, the modal appears again next to the service from Step 1. - it is bug
8. Go to Favorites resource /favourites - only one service is favourite - bug

**Comment by Marcelinna** [comment]
I had problem with add to comparison service when I was in Favorite resource panel
https://www.awesomescreenshot.com/video/3406543?key=8d895e5998f2392e50bdae420b68f337

**Comment by Marcelinna** [comment]
- [ ] DEV
- [ ] QA
Steps 
1. Add service to favorite
2. Add second service to favorite
3. Go to Resource favorite /favourites
4. Remove all services from favorites
5. Come back to /services by using back arrow
6. Services all still marked as added 
https://www.awesomescreenshot.com/video/3408839?key=650bdea790a969578d65051ac09187ba

**Comment by Marcelinna** [comment]
- [ ] DEV
- [ ] QA

Text in modal s different than in the documentation
image
image

**Comment by goreck888** [comment]
>     * [ ]  DEV
> 
>     * [ ]  QA
> 
> 
> I had problem with add to comparison service when I was in Favorite resource panel
> https://www.awesomescreenshot.com/video/3406543?key=8d895e5998f2392e50bdae420b68f337

There are some improvements like visibility of comparison bar with resources included, but there's still something wrong if you try to add a resource to the comparison on the empty list of compared resources.

**Comment by goreck888** [comment]
>     * [ ]  DEV
> 
>     * [ ]  QA
> 
> 
> Steps
> 
>     1. Add service to favorite
> 
>     2. Add second service to favorite
> 
>     3. Go to Resource favorite /favourites
> 
>     4. Remove all services from favorites
> 
>     5. Come back to /services by using back arrow
> 
>     6. Services all still marked as added
>        https://www.awesomescreenshot.com/video/3408839?key=650bdea790a969578d65051ac09187ba

Using back arrow only loads the cached state of a page without any request to the server, so it's impossible to fix it.

**Comment by Marcelinna** [comment]
>     * [x]  DEV
> 
>     * [ ]  QA
> 
> 
> Text in modal s different than in the documentation
> image
> image

Text it is not exactly the same
image

**Comment by goreck888** [comment]
> > ```
> > * [x]  DEV
> > 
> > * [ ]  QA
> > ```
> > 
> > 
> > Text in modal s different than in the documentation
> > image
> > image
> 
> **Text it is not exactly the same**
> image

I agree I forget about comma, but I'm sure there should be "these resources" instead of "this resources"

**Comment by abacz** [comment]
Issues for @jarekzet 
1. Add pb-1 to media-body (service box)
2. Improve popup -> https://invis.io/JXOZKC8WMDU#/447401044_Favourite_Popup
3. I have designed page with no-items status -> https://invis.io/JXOZKC8WMDU#/449980380_Favourite__-_empty

**Comment by abacz** [comment]
@Marcelinna 
Removing favourite items by not-logged-in user shouldn't have affect on logged in favourite list.

**Review by Marcelinna — APPROVED** [review]

**Review by goreck888 — APPROVED** [review]

**Pull request merged**